From a9a2c30bb67cbf1e03d312e02eb659ef9673b123 Mon Sep 17 00:00:00 2001
From: Paul Elder <paul.elder@ideasonboard.com>
Date: Fri, 14 Feb 2025 18:03:53 +0900
Subject: [PATCH] media: imx335: Set vblank immediately

When the vblank v4l2 control is set, it does not get written to the
hardware immediately. It only gets updated when exposure is set. Change
the behavior such that the vblank is written immediately when the
control is set.

Commit c0aa40f45fef ("media: imx335: Set vblank immediately")' upstream

Signed-off-by: Paul Elder <paul.elder@ideasonboard.com>
---
 drivers/media/i2c/imx335.c | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

--- a/drivers/media/i2c/imx335.c
+++ b/drivers/media/i2c/imx335.c
@@ -615,12 +615,12 @@ static int imx335_set_ctrl(struct v4l2_c
 			imx335->vblank,
 			imx335->vblank + imx335->cur_mode->height);
 
-		return __v4l2_ctrl_modify_range(imx335->exp_ctrl,
-						IMX335_EXPOSURE_MIN,
-						imx335->vblank +
-						imx335->cur_mode->height -
-						IMX335_EXPOSURE_OFFSET,
-						1, IMX335_EXPOSURE_DEFAULT);
+		__v4l2_ctrl_modify_range(imx335->exp_ctrl,
+					 IMX335_EXPOSURE_MIN,
+					 imx335->vblank +
+					 imx335->cur_mode->height -
+					 IMX335_EXPOSURE_OFFSET,
+					 1, IMX335_EXPOSURE_DEFAULT);
 	}
 
 	/*
@@ -631,6 +631,13 @@ static int imx335_set_ctrl(struct v4l2_c
 		return 0;
 
 	switch (ctrl->id) {
+	case V4L2_CID_VBLANK:
+		exposure = imx335->exp_ctrl->val;
+		analog_gain = imx335->again_ctrl->val;
+
+		ret = imx335_update_exp_gain(imx335, exposure, analog_gain);
+
+		break;
 	case V4L2_CID_EXPOSURE:
 		exposure = ctrl->val;
 		analog_gain = imx335->again_ctrl->val;
